From 2faa45a8fb9a95d8ef110f2c6ecd91a3c650e2f3 Mon Sep 17 00:00:00 2001
From: Han Xu <han.xu@nxp.com>
Date: Wed, 30 Aug 2017 15:12:35 -0500
Subject: [PATCH 4341/5242] MLK-16329-1: mtd: fsl-flexspi: fix the unalignment
 issue for fspi

commit  721af8b8d589cbf2bafd1ccb1a60a759dda89961 from
https://source.codeaurora.org/external/imx/linux-imx.git

ARM64 platforms may access FSPI from non-64-bit-aligned address which
causes unalignment fault. Fixed the issue for AHB reading.

Signed-off-by: Han Xu <han.xu@nxp.com>
(cherry picked from commit 948c1411230d5e02235fbd429d6ee2e2a2b9adb6)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mtd/spi-nor/fsl-flexspi.c |   19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/drivers/mtd/spi-nor/fsl-flexspi.c b/drivers/mtd/spi-nor/fsl-flexspi.c
index 381894e..4b2267ee 100644
--- a/drivers/mtd/spi-nor/fsl-flexspi.c
+++ b/drivers/mtd/spi-nor/fsl-flexspi.c
@@ -1021,6 +1021,7 @@ static ssize_t fsl_flexspi_read(struct spi_nor *nor, loff_t from,
 		size_t len, u_char *buf)
 {
 	struct fsl_flexspi *flex = nor->priv;
+	int i, j;
 
 	/* if necessary,ioremap buffer before AHB read, */
 	if (!flex->ahb_addr) {
@@ -1053,10 +1054,20 @@ static ssize_t fsl_flexspi_read(struct spi_nor *nor, loff_t from,
 		}
 	}
 
-	/* Read out the data directly from the AHB buffer.*/
-	memcpy(buf,
-	       flex->ahb_addr + flex->chip_base_addr + from - flex->memmap_offs,
-	       len);
+	/* For non-8-byte alignment cases */
+	if (from % 8) {
+		j = 8 - (from & 0x7);
+		for (i = 0; i < j; ++i) {
+			memcpy(buf + i, flex->ahb_addr + flex->chip_base_addr
+			       + from - flex->memmap_offs + i, 1);
+		}
+		memcpy(buf + j, flex->ahb_addr + flex->chip_base_addr + from
+		       - flex->memmap_offs + j, len - j);
+	} else {
+		/* Read out the data directly from the AHB buffer.*/
+		memcpy(buf, flex->ahb_addr + flex->chip_base_addr + from
+		       - flex->memmap_offs, len);
+	}
 
 	return len;
 }
-- 
1.7.9.5

